### PR TITLE
MGMT-8555: Implement UI for ARM

### PR DESCRIPTION
**Pull request by batzionb**
**details**

armcreate

**details after created**

arm3

**hosts**

arm4

**networking**

image

**review and create** 

image

**installation** 

image

**add hosts**

image

### CODE REVIEW

**Comment by nirfarkas** [comment]
LGTM

File: src/common/types/featureSupportLevel.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@ import { FeatureSupportLevel } from '../api/types';
 type ArrayElementType<TWhere> = TWhere extends (infer U)[] ? (U extends object ? U : never) : never;
 
 type Features = Required<ArrayElementType<FeatureSupportLevel['features']>>;
-export type FeatureId = Features['featureId'];
+export type FeatureId = Features['featureId'] | 'OCS' | 'CNV';
```
Comment (jkilzi) [inline src/common/types/featureSupportLevel.ts:7]:
```suggestion
// TODO(jkilzi): remove 'OCS' once https://issues.redhat.com/browse/MGMT-7229 is completed
export type FeatureId = Features['featureId'] | 'OCS' | 'ODF' | 'CNV';
```

File: src/common/components/clusterWizard/networkingSteps/ManagedNetworkingControlGroup.tsx [hunk 2]
```diff
@@ -11,24 +11,19 @@ export const ManagedNetworkingControlGroup = ({
...
+        value={'clusterManaged'}
+        label="Cluster-Managed Networking"
+      />
+      <RadioField
+        name={GROUP_NAME}
+        isDisabled={disabled}
+        value={'userManaged'}
+        label="User-Managed Networking"
+      />
+    </FormGroup>
```
Comment (jkilzi) [inline src/common/components/clusterWizard/networkingSteps/ManagedNetworkingControlGroup.tsx:27]: :+1:

File: src/ocm/components/AddHosts/InventoryAddHost.tsx [hunk 3]
```diff
@@ -8,8 +16,18 @@ import {
...
+  isArmArchitectureCluster,
 } from '../../../common';
 
+const armArchAlert = (
+  <Alert
+    title="Only hosts that have arm64 cpu architecture can be added to this cluster."
+    variant={AlertVariant.info}
+    data-testid="arm-architecture-alert"
+    isInline
+  ></Alert>
```
Comment (jkilzi) [inline src/ocm/components/AddHosts/InventoryAddHost.tsx:28]:
```suggestion
  />
```

File: src/ocm/components/clusterConfiguration/ArmCheckbox.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+          <p>
+            Check this option if you want to use the arm64 CPU architecture instead of the default
+            x86_64 CPU architecture. Please note that some features will not be available.
+          </p>
+        }
+      />
+      <FeatureSupportLevelBadge
+        featureId="ARM64_ARCHITECTURE"
+        openshiftVersion={openshiftVersion}
+      ></FeatureSupportLevelBadge>
```
Comment (jkilzi) [inline src/ocm/components/clusterConfiguration/ArmCheckbox.tsx:33]:
```suggestion
      />
```

File: src/ocm/services/apis/ClustersAPI.ts [hunk 5]
```diff
@@ -118,6 +119,9 @@ const ClustersAPI = {
   },
 
   register(params: ClusterCreateParams) {
+    if (params.cpuArchitecture === CpuArchitecture.ARM) {
+      params.userManagedNetworking = true;
+    }
```
Comment (jkilzi) [inline src/ocm/services/apis/ClustersAPI.ts:124]: This logic doesn't belong here, the ClustersAPI is meant to be a simple REST client.
Please perform this logic [inside the caller](https://github.com/openshift-assisted/assisted-ui-lib/blob/e1d13b7d3e82dd933d82090afec479e957e151b3/src/ocm/services/ClusterDetailsService.ts#L9) of this function.

File: src/ocm/components/featureSupportLevels/utils.tsx [hunk 6]
```diff
@@ -53,3 +60,87 @@ export const getLimitedFeatureSupportLevels = (
...
+  if (!isSupported) {
+    return 'Single-Node OpenShift is not supported in this OpenShift version';
+  }
+  return '';
+};
+
+const getArmDisabledReason = (
+  cluster: Cluster | undefined,
+  version: string,
+  versions: OpenshiftVersionOptionType[],
```
Comment (jkilzi) [inline src/ocm/components/featureSupportLevels/utils.tsx:83]: What is the difference between `version` and `versions`?
Can we give `version` a name that aligns with its context?

File: src/ocm/components/featureSupportLevels/utils.tsx [hunk 7]
```diff
@@ -53,3 +60,87 @@ export const getLimitedFeatureSupportLevels = (
     return undefined;
   }
 };
+
+const isArmSupported = (version: string, versions: OpenshiftVersionOptionType[]) => {
```
Comment (jkilzi) [inline src/ocm/components/featureSupportLevels/utils.tsx:64]: What is the difference between `version` and `versions`?
Can we give `version` a name that aligns with its context?

File: src/ocm/components/featureSupportLevels/utils.tsx [hunk 8]
```diff
@@ -53,3 +60,87 @@ export const getLimitedFeatureSupportLevels = (
...
+    }
+    case 'ARM64_ARCHITECTURE': {
+      return getArmDisabledReason(cluster, version, openshiftVersions);
+    }
+    case 'CNV': {
+      return getCnvDisabledReason(cluster);
+    }
+    case 'OCS': {
+      return getOcsDisabledReason(cluster);
+    }
```
Comment (jkilzi) [inline src/ocm/components/featureSupportLevels/utils.tsx:128]:
```suggestion
    case 'ODF':
    case 'OCS': {
      return getOcsDisabledReason(cluster);
    }
```

File: src/ocm/components/clusterWizard/ClusterDetailsForm.tsx [hunk 9]
```diff
@@ -50,7 +51,7 @@ const ClusterDetailsForm: React.FC<ClusterDetailsFormProps> = (props) => {
   } = props;
 
   const featureSupportLevels = React.useContext(FeatureSupportLevelContext);
-  const handleSubmit = async (values: ClusterDetailsValues) => {
+  const handleSubmit = async (values: OcmClusterDetailsValues) => {
```
Comment (jkilzi) [inline src/ocm/components/clusterWizard/ClusterDetailsForm.tsx:54]: Please wrap it inside `React.useCallback`

**Review by jkilzi — CHANGES_REQUESTED** [review]
Batz, large PR but in general, Kudos! and nicely done!
Please go over my comments, let me know your thoughts about them :100:

File: src/ocm/components/clusterWizard/ClusterDetailsForm.tsx (hunk 9 above)
Comment (batzionb) [reply src/ocm/components/clusterWizard/ClusterDetailsForm.tsx:54]: you mean wrap handleSubmit?

Comment (jkilzi) [reply src/ocm/components/clusterWizard/ClusterDetailsForm.tsx:54]: yes

File: src/common/types/featureSupportLevel.ts (hunk 1 above)
Comment (batzionb) [reply src/common/types/featureSupportLevel.ts:7]: since 'OCS' feature id isn't affected by the backend I'll change it to 'ODF' and not add both

**Comment by jkilzi** [comment]
@batzionb please update your branch

File: src/ocm/components/AddHosts/InventoryAddHost.tsx (hunk 3 above)
Comment (jkilzi) [reply src/ocm/components/AddHosts/InventoryAddHost.tsx:28]: what about this?

File: src/ocm/components/AddHosts/InventoryAddHost.tsx [hunk 10]
```diff
@@ -19,28 +37,38 @@ const InventoryAddHosts: React.FC = () => {
...
             cluster={cluster}
-            idPrefix="bare-metal-inventory-add-host"
+            HostRequirementsContent={AddHostRequirementsContent}
+            setDiscoveryHintModalOpen={setDiscoveryHintModalOpen}
           />
-        </Text>
-        <InformationAndAlerts
+        </TextContent>
+      </StackItem>
+      <StackItem>{isArmArchitecture(cluster) && armArchAlert}</StackItem>
```
Comment (jkilzi) [inline src/ocm/components/AddHosts/InventoryAddHost.tsx:58]: The StackItem is redundant when `isArmArchitecture(cluster)` is false 
```suggestion
      {isArmArchitecture(cluster) && <StackItem>armArchAlert</StackItem>}
```

File: src/ocm/components/clusterConfiguration/ArmCheckbox.tsx (hunk 4 above)
Comment (jkilzi) [reply src/ocm/components/clusterConfiguration/ArmCheckbox.tsx:33]: What about this?

**Review by jkilzi — CHANGES_REQUESTED** [review]
Batz, there are still some minor changes needed...

File: src/common/components/clusterConfiguration/NetworkConfiguration.tsx [hunk 11]
```diff
@@ -15,7 +15,7 @@ import { ClusterDefaultConfig } from '../../api';
 import { isSingleNodeCluster } from '../clusters';
 import { NO_SUBNET_SET } from '../../config';
 import { isAdvNetworkConf } from './utils';
-import { getLimitedFeatureSupportLevels } from '../../../ocm/components/featureSupportLevels/utils';
+import { getLimitedFeatureSupportLevels } from '../../../ocm/components/featureSupportLevels/utils'; //TODO(brotman): move OCM dependency to common
```
Comment (jkilzi) [inline src/common/components/clusterConfiguration/NetworkConfiguration.tsx:18]: `common` depending on `ocm` :(
Is this going to stay here forever?

**Review by jkilzi — APPROVED** [review]

**Pull request merged**